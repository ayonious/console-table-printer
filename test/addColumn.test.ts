import { Table } from '../index';

describe('Testing adding columns', () => {
  it('should allow add new columns in a chain way', () => {
    const p = new Table()
      .addColumn('foo')
      .addColumns(['bar'])
      .addRow({ foo: '1' })
      .addRows([{ bar: '2' }]);

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should allow adding a single column', () => {
    const p = new Table().addColumn('singleColumn');
    p.addRow({ singleColumn: 'value' });
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should allow adding multiple columns', () => {
    const p = new Table().addColumns(['col1', 'col2']);
    p.addRow({ col1: 'value1', col2: 'value2' });
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should allow adding columns with different alignments', () => {
    const p = new Table()
      .addColumn({ name: 'leftAligned', alignment: 'left' })
      .addColumn({ name: 'rightAligned', alignment: 'right' });
    p.addRow({ leftAligned: 'left', rightAligned: 'right' });
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should allow adding columns with colors', () => {
    const p = new Table()
      .addColumn({ name: 'coloredColumn', color: 'red' });
    p.addRow({ coloredColumn: 'redValue' });
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should allow adding a complex column object with chaining', () => {
    const p = new Table()
      .addColumn({ name: 'complexColumn', alignment: 'center', color: 'blue', title: 'Complex Column' })
      .addColumn('simpleColumn')
      .addRow({ complexColumn: 'complexValue', simpleColumn: 'simpleValue' });
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should correctly set internal column properties', () => {
    const p = new Table()
      .addColumn({ 
        name: 'testCol', 
        alignment: 'center', 
        color: 'blue',
        maxLen: 20,
        title: 'Test Column'
      });

    expect(p.table.columns[0].name).toBe('testCol');
    expect(p.table.columns[0].alignment).toBe('center');
    expect(p.table.columns[0].color).toBe('blue');
    expect(p.table.columns[0].maxLen).toBe(20);
    expect(p.table.columns[0].title).toBe('Test Column');
  });

  it('should allow adding columns after rows are added', () => {
    const p = new Table()
      .addColumn('col1')
      .addRow({ col1: 'value1' })
      .addColumn('col2');

    expect(p.table.columns.length).toBe(2);
    expect(p.table.columns[1].name).toBe('col2');
    p.addRow({ col1: 'value2', col2: 'added' });
    
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle column name overrides correctly', () => {
    const p = new Table()
      .addColumn('col1')
      .addColumn({ name: 'col1', alignment: 'right' }); // Override existing column

    expect(p.table.columns.length).toBe(1);
    expect(p.table.columns[0].alignment).toBe('right');
    
    p.addRow({ col1: 'test' });
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should preserve column order when adding multiple columns', () => {
    const p = new Table()
      .addColumns(['col1', 'col2', 'col3']);

    expect(p.table.columns[0].name).toBe('col1');
    expect(p.table.columns[1].name).toBe('col2');
    expect(p.table.columns[2].name).toBe('col3');

    p.addRows([
      { col1: '1', col2: '2', col3: '3' },
      { col1: '4', col2: '5', col3: '6' }
    ]);
    
    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });
});
