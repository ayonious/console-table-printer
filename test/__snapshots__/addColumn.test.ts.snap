// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing adding columns should allow add new columns in a chain way 1`] = `
"┌─────┬─────┐
│[37m [0m[01mfoo[0m │[37m [0m[01mbar[0m │
├─────┼─────┤
│[37m [0m[37m  1[0m │[37m [0m[37m   [0m │
│[37m [0m[37m   [0m │[37m [0m[37m  2[0m │
└─────┴─────┘"
`;

exports[`Testing adding columns should allow adding a complex column object with chaining 1`] = `
"┌────────────────┬──────────────┐
│[37m [0m[01mComplex Column[0m │[37m [0m[01msimpleColumn[0m │
├────────────────┼──────────────┤
│[37m [0m[34m complexValue [0m │[37m [0m[37m simpleValue[0m │
└────────────────┴──────────────┘"
`;

exports[`Testing adding columns should allow adding a single column 1`] = `
"┌──────────────┐
│[37m [0m[01msingleColumn[0m │
├──────────────┤
│[37m [0m[37m       value[0m │
└──────────────┘"
`;

exports[`Testing adding columns should allow adding columns after rows are added 1`] = `
"┌────────┬───────┐
│[37m [0m[01m  col1[0m │[37m [0m[01m col2[0m │
├────────┼───────┤
│[37m [0m[37mvalue1[0m │[37m [0m[37m     [0m │
│[37m [0m[37mvalue2[0m │[37m [0m[37madded[0m │
└────────┴───────┘"
`;

exports[`Testing adding columns should allow adding columns with colors 1`] = `
"┌───────────────┐
│[37m [0m[01mcoloredColumn[0m │
├───────────────┤
│[37m [0m[31m     redValue[0m │
└───────────────┘"
`;

exports[`Testing adding columns should allow adding columns with different alignments 1`] = `
"┌─────────────┬──────────────┐
│[37m [0m[01mleftAligned[0m │[37m [0m[01mrightAligned[0m │
├─────────────┼──────────────┤
│[37m [0m[37mleft       [0m │[37m [0m[37m       right[0m │
└─────────────┴──────────────┘"
`;

exports[`Testing adding columns should allow adding multiple columns 1`] = `
"┌────────┬────────┐
│[37m [0m[01m  col1[0m │[37m [0m[01m  col2[0m │
├────────┼────────┤
│[37m [0m[37mvalue1[0m │[37m [0m[37mvalue2[0m │
└────────┴────────┘"
`;

exports[`Testing adding columns should preserve column order when adding multiple columns 1`] = `
"┌──────┬──────┬──────┐
│[37m [0m[01mcol1[0m │[37m [0m[01mcol2[0m │[37m [0m[01mcol3[0m │
├──────┼──────┼──────┤
│[37m [0m[37m   1[0m │[37m [0m[37m   2[0m │[37m [0m[37m   3[0m │
│[37m [0m[37m   4[0m │[37m [0m[37m   5[0m │[37m [0m[37m   6[0m │
└──────┴──────┴──────┘"
`;
